new lint: `eager_transmute`

y21 opened the pull request:
A small but still hopefully useful lint that looks for patterns such as `(x < 5).then_some(transmute(x))`.
This is almost certainly wrong because it evaluates the transmute eagerly and can lead to surprises such as the check being completely removed and always evaluating to `Some` no matter what `x` is (it is UB after all when the integer is not a valid bitpattern for the transmuted-to type). [Example](https://godbolt.org/z/xoY34fPzh).
The user most likely meant to use `then` instead.

I can't remember where I saw this but this is inspired by a real bug that happened in practice.

This could probably be a correctness lint?

changelog: new lint: [`eager_int_transmute`]

KisaragiEffective commented on clippy_lints/src/transmute/eager_int_transmute.rs:
```diff
@@ -0,0 +1,39 @@
+            expr.span,
+            "this transmute is evaluated eagerly, even if the condition is false",
+            Some(path.ident.span),
+            "consider using `then` instead, which evaluates it only if the condition is true",
```
(Sorry if this is wrong)
I think suggesting this symbol requires MSRV-gating because [it](https://doc.rust-lang.org/stable/std/primitive.bool.html#method.then) is stabilized in 1.50?

llogiq commented:
I'd leave it as `suspicious`  until we add a check of the result type. Also we should have a test that confirms the risk of false positive, as in `let x: u64 = (foo < 16).then_some(std::mem::transmute(foo));`. Otherwise this looks good, r=me with the test added.

y21 commented:
> Also we should have a test that confirms the risk of false positive, as in `let x: u64 = (foo < 16).then_some(std::mem::transmute(foo));`.

What's the false positive here? Transmuting to a type with the same validity range (`u64`->`u64`)?

llogiq commented:
That or a higher validity range, yes.

y21 commented:
Added some more logic + tests to only lint if the target type has a smaller validity range and also renamed to `eager_transmute` since it might be possible to make the lint more general

llogiq commented:
I think the way the lint is now, we should actually put this in `correctness`, because I don't see any possibility of false positives and the lint rules out UB.

llogiq commented:
Thank you!

@bors r+

bors commented:
:pushpin: Commit 08d8ca9edd7a83f15a27012391fc9760ed30e408 has been approved by `llogiq`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/clippy) for this repository.

<!-- @bors r=llogiq 08d8ca9edd7a83f15a27012391fc9760ed30e408 -->
<!-- homu: {"type":"Approved","sha":"08d8ca9edd7a83f15a27012391fc9760ed30e408","approver":"llogiq","queue":"https://bors.rust-lang.org/homu/queue/clippy"} -->

bors commented:
:hourglass: Testing commit 08d8ca9edd7a83f15a27012391fc9760ed30e408 with merge c689d32a908377bdc92925275ef258060598e6a0...
<!-- homu: {"type":"BuildStarted","head_sha":"08d8ca9edd7a83f15a27012391fc9760ed30e408","merge_sha":"c689d32a908377bdc92925275ef258060598e6a0"} -->

bors commented:
:sunny: Test successful - [checks-action_dev_test](https://github.com/rust-lang/rust-clippy/actions/runs/7339593048/job/19984112781), [checks-action_remark_test](https://github.com/rust-lang/rust-clippy/actions/runs/7339593065/job/19984118559), [checks-action_test](https://github.com/rust-lang/rust-clippy/actions/runs/7339593055/job/19984227824)
Approved by: llogiq
Pushing c689d32a908377bdc92925275ef258060598e6a0 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"llogiq","base_ref":"master","builders":{"checks-action_test":"https://github.com/rust-lang/rust-clippy/actions/runs/7339593055/job/19984227824","checks-action_dev_test":"https://github.com/rust-lang/rust-clippy/actions/runs/7339593048/job/19984112781","checks-action_remark_test":"https://github.com/rust-lang/rust-clippy/actions/runs/7339593065/job/19984118559"},"merge_sha":"c689d32a908377bdc92925275ef258060598e6a0"} -->
